app/lib/payment_gateways/stripe_crypt.rb

```diff
@@ -27,31 +27,35 @@ def create_stripe_setup_intent
       setup_intent_params = {
         payment_method_types: ['card'],
         usage: 'off_session',
-        customer: find_or_create_customer.id
+        customer: customer.id
       }
       Stripe::SetupIntent.create(setup_intent_params, api_key)
     end
 
+    def customer
+      @customer ||= find_or_create_customer
+    end
+
     private
 
     delegate :payment_detail, to: :account
 
     def find_or_create_customer
       customer_id = payment_detail.credit_card_auth_code
-      customer_id.present? ? Stripe::Customer.retrieve(customer_id, api_key) : create_customer
+      return create_customer if customer_id.blank?
+
+      customer = Stripe::Customer.retrieve(customer_id, api_key)
+      customer.deleted? ? create_customer : customer
```

Comment:
Ok. I see you've more or less changed it back. But is `customer.deleted?` equivalent to `customer.id == nil`? One thing is having been deleted, another perhaps it is never having been created and therefore not retrieved in the first place.

Reply:
>When requesting the ID of a customer that has been deleted, a subset of the customer’s information will be returned, including a deleted property, which will be true.

I think both should be checked. We don't know what "subset" means here but it could even include an ID. Maybe the proper way is "listing" customers (filtered by email?) and check whether it exists there.

Reply:
I think that means that the customer was deleted in Stripe, which is safe to check. Because now we do not know if the customer was deleted or not in Stripe for the current Account.

Reply:
If the customer has never been created, it won't return nil, it will raise an error

Reply:
I will ensure what happens and cover that case

Reply:
![image](https://user-images.githubusercontent.com/11318903/105726411-55f02a00-5f2a-11eb-9dcf-b20a2bf0c947.png)

This is what I meant... If the customer exists but was once deleted vs when the customer has never existed

Reply:
@guicassolato The `new` builds the empty object (having its `id` present though), but the `refresh` does the request to the internet, and that's where the error is raised in that scenario
https://github.com/stripe/stripe-ruby/blob/9787913b355049af915cf075b29666e55c5ab7f2/lib/stripe/api_resource.rb#L101-L106

Reply:
Basically the options are:
- Returning `nil` can NEVER happen.
- Return an object with `customer.id == nil` can NEVER happen
- Returning an object with the entire customer info with `deleted?` responding to `true` is when the customer has been created in the past in Stripe but has been deleted there.
- Returning an object with the entire customer info with `deleted?` responding to `false` is when the customer was once created in Stripe and has never been deleted there.
- Raising `Stripe::InvalidRequestError` happens when the customer has never been created in Stripe for this 'tenant'.

Reply:
Another possible error raised is not having connectivity to Stripe btw 😉

Reply:
`Stripe::InvalidRequestError` is not something that I ever expect to happen, because if it has never been created, then how on earth that ID ended in our DB as a valid customer ID?
Anyway, if you want, I can rescue that error and report to Bugsnag just in case.

Reply:
Ok. Good enough then. Thanks.